README.md
# Banan-Schedule-Automation
Automate banana schedule from my email to my calendar

I get banan republic email every week and I have to manually put it into my google calendar. then I realize "wait a minute, I can code". 

Objective:
Automate the process to add my Banana Republic work schedule to my calendar

Tasks
* get work schedule from my gmail
  * technical spike on gmail API
    * following steps from to get email
      https://developers.google.com/gmail/api/quickstart/nodejs
    
  * access gmail
  * find the right email
  * get thatemail
* parse work the email to google calendar format
* add it to google calendar

12.04.2020
We will need a server to make the API call to gmail and calendar. that is what server.js is for.

12.18.2020
This is how far I got: 
step 1 "npm i"
step 2 "node ." to run index.js
step 3 "the terminal would say "Authorize this app by visiting this url: https://... Enter the code from that page here: "
step 4 Go to the websire, Chrom would tell you not to be go ahead anyway
step 5 Get the code, base it in the terminal
step 6 It would generate token.json with and list all the token you have in the termnial
this is as far as I go for now

12.28.2020
How to seperat the two types:

opt1: make 2 api call, one to get all the weekly schedule email
                       one to get all the monthly schedule email
      con: * might take longer for 2 round trip
           * might be more expensive too
      pro: * there are clear sperateions, more organized
Opt2: make one call and loop through to array to speate them 
      con: * messier potentially
      pro: only one call might be faster
  note: the time tride of are probably negligible

Decision:
  use snippet becuase it is easier for now
  if I use suject then I have to look through the payload headers
  I thought about which would be more resilient for changes but if they change the email templates there is a big chance that my code would not work anyway.

Idea for implamentation:
  If there is a email that my code cann't read, the return the url to that email and or the email object for me to play with it.

The month long one snippet:
"Schedule Changes/Changements d&#39;horaire/Cambios de horario DATE/L&#39;DATE/LA FECHA SCHEDULE/LE PLANNING/EL HORARIO UPDATED/MIS À JOUR/ACTUALIZADO DEPARTMENT/DÉPARTEMENT/DEPARTAMENTO ACTIVITY/"

The week long one snippet:
Sunday/Dimanche/Domingo, 11/22 01:00PM - 10:00PM WRK OUT_SALES, 01049_OUTL, 01049_OUTL_SALES_FLOOR Monday/Lundi/Lunes, 11/23 Off/Congé/Inactivo Tuesday/Mardi/Martes, 11/24 Off/Congé/Inactivo Wednesday/

12.28.2020
tested speed for getting and parshing the email array in a loop and a map

loop is around 390 ms
map is around 450 ms

12.28.2020
shit, I need the date that email is sent too to see which email is the latest change

** Leason Learned **

You don't refind the app as you go

you need to run all the way through first to see all the problem so you can evaluate the sevarity of the issues. 

if you fix and refine as you go, you will spend a lot of time on something that becamse irrelivent later down the road.

12.28.2020
shit, what if you had a shift on a date then a schdule changes and you don't have the shift on that date no more

right now I am using the list of shift from gmail and see if they need to be added or updated. this will cause problem becuase it would not remove the shift that's taken away in the later changes

but if there is a shift that is taken away, I would want to be alerted for that

in fact, I would like to be alerted of any changes to my schudule

01.13.2021

1304: organize example data with test into a folder and see if that I can still run test with "npm run test"

01.17.2021

I need to have a grand plan. I need draw a plan for how each function is invoked it to. Step by step 

01.18.2021

Grand Plan: 

* Gmail Side 
This app will get the last X email from Ray's gmail sent by "no-reply-ams@infor.com"

It will parse the email depending on the email template

It will compile a master shift schedule with all the dates included in the emails regardless the email indicates there is shift for that day

It will indicate on the master shift schedule whether the user have shifts for the date range that the emails covered

* Calendar Side
It will takes in the master shift scheudule 

loop through all the dates

  get shift event for that date if any

  check if it matech the shift in the master schedule{
    if match
      do nothing
    if don't match{
      make changes to the calendare to match master shift schedule{
        if(master schulde have shift for that date){
          if(calendar have a shift on that date as well){
            modify the shift to match the master schedule
          }else if the calendar doesn't have the shift {
            add the shift from master shift schedule to calendar
          }
        } doesn't {
          alert user
          rm the shift
        }
      }
    }
    log update in the event description if the event exist
  }

  01.19.2021
  Working on emailParser_UTL.js 
  out of the 4 functions in that file I got first 2 to work yesterday. 
  
  Today I am working on emailContentParser_htmlTemplate function
    I got to read the html template with the next of "node-html-parser"
    I got to trivers the html table columns





CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.2] - 2021-01-19
### Added
- added emailParser_UTL for all email parsing functions
- added htmlEmail.html to help visualize html template
- added node-html-parser npm package to help iterate through html parser
- added htmlEmail_structure.txt to help visualize html template structure from node-html-parser prespective

## [0.0.1] - 2021-01-18
### Added
- Change Log




### Below are just Keep a Changelog template

## [Unreleased]

## [1.0.0] - 2017-06-20
### Added
- New visual identity by [@tylerfortune8](https://github.com/tylerfortune8).
- Version navigation.
- Links to latest released version in previous versions.
- "Why keep a changelog?" section.
- "Who needs a changelog?" section.
- "How do I make a changelog?" section.
- "Frequently Asked Questions" section.
- New "Guiding Principles" sub-section to "How do I make a changelog?".
- Simplified and Traditional Chinese translations from [@tianshuo](https://github.com/tianshuo).
- German translation from [@mpbzh](https://github.com/mpbzh) & [@Art4](https://github.com/Art4).
- Italian translation from [@azkidenz](https://github.com/azkidenz).
- Swedish translation from [@magol](https://github.com/magol).
- Turkish translation from [@karalamalar](https://github.com/karalamalar).
- French translation from [@zapashcanon](https://github.com/zapashcanon).
- Brazilian Portugese translation from [@Webysther](https://github.com/Webysther).
- Polish translation from [@amielucha](https://github.com/amielucha) & [@m-aciek](https://github.com/m-aciek).
- Russian translation from [@aishek](https://github.com/aishek).
- Czech translation from [@h4vry](https://github.com/h4vry).
- Slovak translation from [@jkostolansky](https://github.com/jkostolansky).
- Korean translation from [@pierceh89](https://github.com/pierceh89).
- Croatian translation from [@porx](https://github.com/porx).
- Persian translation from [@Hameds](https://github.com/Hameds).
- Ukrainian translation from [@osadchyi-s](https://github.com/osadchyi-s).

### Changed
- Start using "changelog" over "change log" since it's the common usage.
- Start versioning based on the current English version at 0.3.0 to help
translation authors keep things up-to-date.
- Rewrite "What makes unicorns cry?" section.
- Rewrite "Ignoring Deprecations" sub-section to clarify the ideal
  scenario.
- Improve "Commit log diffs" sub-section to further argument against
  them.
- Merge "Why can’t people just use a git log diff?" with "Commit log
  diffs"
- Fix typos in Simplified Chinese and Traditional Chinese translations.
- Fix typos in Brazilian Portuguese translation.
- Fix typos in Turkish translation.
- Fix typos in Czech translation.
- Fix typos in Swedish translation.
- Improve phrasing in French translation.
- Fix phrasing and spelling in German translation.

### Removed
- Section about "changelog" vs "CHANGELOG".

## [0.3.0] - 2015-12-03
### Added
- RU translation from [@aishek](https://github.com/aishek).
- pt-BR translation from [@tallesl](https://github.com/tallesl).
- es-ES translation from [@ZeliosAriex](https://github.com/ZeliosAriex).

## [0.2.0] - 2015-10-06
### Changed
- Remove exclusionary mentions of "open source" since this project can
benefit both "open" and "closed" source projects equally.

## [0.1.0] - 2015-10-06
### Added
- Answer "Should you ever rewrite a change log?".

### Changed
- Improve argument against commit logs.
- Start following [SemVer](https://semver.org) properly.

## [0.0.8] - 2015-02-17
### Changed
- Update year to match in every README example.
- Reluctantly stop making fun of Brits only, since most of the world
  writes dates in a strange way.

### Fixed
- Fix typos in recent README changes.
- Update outdated unreleased diff link.

## [0.0.7] - 2015-02-16
### Added
- Link, and make it obvious that date format is ISO 8601.

### Changed
- Clarified the section on "Is there a standard change log format?".

### Fixed
- Fix Markdown links to tag comparison URL with footnote-style links.

## [0.0.6] - 2014-12-12
### Added
- README section on "yanked" releases.

## [0.0.5] - 2014-08-09
### Added
- Markdown links to version tags on release headings.
- Unreleased section to gather unreleased changes and encourage note
keeping prior to releases.

## [0.0.4] - 2014-08-09
### Added
- Better explanation of the difference between the file ("CHANGELOG")
and its function "the change log".

### Changed
- Refer to a "change log" instead of a "CHANGELOG" throughout the site
to differentiate between the file and the purpose of the file — the
logging of changes.

### Removed
- Remove empty sections from CHANGELOG, they occupy too much space and
create too much noise in the file. People will have to assume that the
missing sections were intentionally left out because they contained no
notable changes.

## [0.0.3] - 2014-08-09
### Added
- "Why should I care?" section mentioning The Changelog podcast.

## [0.0.2] - 2014-07-10
### Added
- Explanation of the recommended reverse chronological release ordering.

## [0.0.1] - 2014-05-31
### Added
- This CHANGELOG file to hopefully serve as an evolving example of a
  standardized open source project CHANGELOG.
- CNAME file to enable GitHub Pages custom domain
- README now contains answers to common questions about CHANGELOGs
- Good examples and basic guidelines, including proper date formatting.
- Counter-examples: "What makes unicorns cry?"

[Unreleased]: https://github.com/olivierlacan/keep-a-changelog/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/olivierlacan/keep-a-changelog/compare/v0.3.0...v1.0.0
[0.3.0]: https://github.com/olivierlacan/keep-a-changelog/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/olivierlacan/keep-a-changelog/compare/v0.1.0...v0.2.0

index.js
const { getBREventList,
        deleteEvent,
        addEvent,
        patch,
        getBREventByDate,
        initCalendar } = require('./calendarAPIv2.js');
const { getMessageList, 
        getMessageContent, 
        parseMessageContent, 
        initGmail 
      } = require('./gmailAPIv2.js');
const { getEmailMetaInfo, 
        emailParser_base64,
        emailContnetParser_htmlTemplate,
        emailContnetParser_planTextTemplate,
      } = require('./emailParser_UTL.js');
// const { email_Id_List } = require('./testEmailId.js')

const get_Email_Ids = async(parameters) => {
  const gmail = await initGmail()
  return new Promise((resolve, reject) => {
    gmail.users.messages.list(parameters, (err, res) => {
      if (err) {return reject(`Error in get_Email_Ids funcrtion.\nParameters:\n${parameters}\n${err}`)};
      const { messages } = res.data;
      resolve(messages)
    })
  })
}


const get_Email_Content = async(parameters) => {
  const gmail = await initGmail()
  return new Promise((resolve, reject) => {
    gmail.users.messages.get(parameters, (err, res) => {
      if (err) {return reject(`Error in get_Email_Content function\nParameters:\n${parameters}\n${err}`)};
      const { data } = res;
      resolve(data)
    })
  })
}

const parseEmail = (emailObj) => {
  if(!emailObj) {
    throw Error(`"parseEmail" function expect input to be valid emailObj but recieved ${emailObj}`)
  }
  const { id, payload } = emailObj
  const metaData = {
    id,
    contentType: payload.mimeType,
  }
  const reducer = (sto, cur) => {
    if(!sto){
      sto = {}
    }
    const reqElements = ['From', 'Subject', 'Date', 'Content-Type'];
    reqElements.forEach((ele) => {
      if(cur.name === ele){
        sto
      }
    })
  }
  payload.headers.reduce()
  
}

const auto_update_shift = async() => {
  const calendar = initCalendar();
  const gmail = initGmail();
  const emailListParam = {
    userId: 'me',
    includeSpamTrash: false,
    labelIds: 'INBOX',
    maxResults: 5,
    q: 'from:no-reply-ams@infor.com'
  };
  const email_Id_List = await get_Email_Ids(emailListParam)

  //*** HAVE EMAIL ID ARRAY */

  const email_Content_List = await Promise.all(
    email_Id_List.map((ele) => (getMessageContent(ele.id)))
  )
  //*** HAVE EMAIL CONTENT ARRAY */
  let masterSchedule = {};

  email_Content_List.map((email) => {
    const emailType = email.payload.mimeType;
    const metaInfo = getEmailMetaInfo(email);
    const emailBody = emailParser_base64(email.payload.body.data);
    let shiftsInEmail = {};
    if(emailType === "text/html"){
      shiftsInEmail = emailContnetParser_htmlTemplate(emailBody, metaInfo);
    }else if(emailType === "text/plain"){
      // let year = new Date(metaInfo.dateRecieved)
      // year = JSON.stringify(year.getFullYear());
      shiftsInEmail = emailContnetParser_planTextTemplate(emailBody, metaInfo)
    }else{
      //alert me
      throw Error("email_Content_List mapping problem, see index.js line 86 to 99");
    }
    masterSchedule = {
      ...masterSchedule,
      ...shiftsInEmail,
    }
  })

  // console.log(masterSchedule)
  return masterSchedule
}

// auto_update_shift()
// .then(console.log)
// .catch(console.log)

const { exMasterSchedule } = require('./devHelper/exMasterSchedule.js');

const checkCalHasShift = async (shiftInfo) => {
  const { date, schedule } = shiftInfo;
  const oneDateRane = (date) => {
    const stTime = `${date} 00:00:00`;
    const endTime = `${date} 23:59:59`;
    return [ new Date(stTime), new Date(endTime)];
  }
  const [ stDate, endDate ] = oneDateRane(date);
  const calShiftThatDate = await getBREventByDate(stDate, endDate)
  return calShiftThatDate
};

const testShift = {
  date: '01/26/2021',
  schedule: [ '03:00 PM', '07:00 PM' ],
  updated: true,
  department: '01049_OUTL_SALES_FLOOR',
  activity: 'WRK',
  store: '01049_OUTL',
  job: 'OUT_SALES',
  srcEmailInfo: {
    id: '176dfab727d9a969',
    dateRecieved: 'Fri, 8 Jan 2021 01:45:23 +0000',
    from: 'no-reply-ams@infor.com',
    subject: "Schedule Change Alert/Alerte de changement d'horaire/Programar alerta de cambio",
    content_type: 'text/html; charset=ISO-8859-1'
  }
}

const testPatch_Shift = {
  date: '01/26/2021',
  schedule: [ '03:00 PM', '07:00 PM' ],
  updated: true,
  department: '01049_OUTL_SALES_FLOOR',
  activity: 'WRK',
  store: '01049_OUTL',
  job: 'OUT_SALES',
  srcEmailInfo: {
    id: '176dfab727d9a969',
    dateRecieved: 'Fri, 8 Jan 2021 01:45:23 +0000',
    from: 'no-reply-ams@infor.com',
    subject: "Schedule Change Alert/Alerte de changement d'horaire/Programar alerta de cambio",
    content_type: 'text/html; charset=ISO-8859-1'
  }
}

// checkCalHasShift(testShift)
// .then(console.log)
// .catch(console.log)
const updateShiftHist = (calShift, shiftInfo_mst, notes) => {
  console.log("shiftInfo_mst in progress...")
};
const addShift = async(shiftInfo_mst) => {
  let newEvent = null;
  const shiftEvent = generateShiftEvent(shiftInfo_mst)

  try{
    newEvent = await addEvent(shiftEvent);
    // console.log(`Event ${newEven.id} has been created`);
    // console.log(`Event url:${newEven.url}`);

  }catch (error){
    console.error(`Error occured in adding shifft for ${shiftInfo_mst.date} from ${shiftInfo_mst.srcEmailInfo.dateRecieved} email ${error}`);
  }
  return newEvent
}

// addShift(testShift)
// .then(console.log)
// .catch(console.log)

const changeShiftTime = (calShift, shiftInfo_mst) => {

}

const testChangeShift = async() => {
  const {newShiftId, newShiftUrl} = await addShift(testShift);
  const calSH = await getBREventByDate(new Date('01/26/2021 00:00:00'), new Date('01/26/2021 23:59:59'));
  changeShiftTime(calShift, )
}

testChangeShift()
.then(console.log)
.catch(console.log)


function generateShiftEvent(shiftInfo_mst) {
  const dateTimeStr = (date, time) => {
    return new Date(`${date} ${time}`).toISOString();
  }
  const stDate = dateTimeStr(shiftInfo_mst.date, shiftInfo_mst.schedule[0])
  const endDate = dateTimeStr(shiftInfo_mst.date, shiftInfo_mst.schedule[1])
  const calEvenResource = {
    'summary': 'test',
    'location': '2990 Livermore Outlets Dr SUITE 2990, Livermore, CA 94551',
    'description': 'description secription',
    'start': {
      'dateTime': stDate,
      'timeZone': 'America/Los_Angeles',
    },
    'end': {
      'dateTime': endDate,
      'timeZone': 'America/Los_Angeles',
    },
    'attendees': [
    ],
    'reminders': {
      'useDefault': false,
      'overrides': [
        {'method': 'popup', 'minutes': 10},
      ],
    },
  };
  return calEvenResource;
};


const addShiftsToCalendar = async(masterSchedule) => {
  //for every shift checking if there is already a shift that date 

  for(let shiftDate in masterSchedule){
    /** BLOCKing ASYNC PROBLEM!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! */

    const shiftInfo_mst = masterSchedule[shiftDate];
    //checkShift return bootlean, yes: return shift, no return false
    const calShift = await checkCalHasShift(shiftInfo_mst);
    const calHasSameDateShift = calShift.length;
    const hasShiftAtSameTime = false;
    if(calHasSameDateShift){
      if(hasShiftAtSameTime){
        //Patch
        updateShiftHist(calShift, shiftInfo_mst, "Confirmed, No Changes");
        //event descrip: "updated 01/28/2021 0902, emailId:blah blach"

      }else{
        //Patch
        //change shiftTime to match masterSchedule
        changeShiftTime(calShift, shiftInfo_mst);
        //event description: "changed from 0400 ro 0800" (lastModdate) to ""0500 to 1100"
      }
    }else{
      addShift(shiftInfo_mst);
      //Add
      //addShiftToCal()
    }
  }

  // return Object.keys(masterSchedule)
}

// addShiftsToCalendar(exMasterSchedule)
// .then(console.log)
// .catch(console.log)



// get_Email_Content({
//   userId: 'me',
//   id: '176794e0d843c7c5',
//   format: "full",
// })
// .then(console.log)
// .catch(console.log)